include/sys/onoff.h

```diff
@@ -0,0 +1,533 @@
+ * the client.
+ */
+enum onoff_client_flags {
+	/* Indicates operations are to proceed only if the initiating
+	 * function can be executed in a synchronous path.
+	 *
+	 * I.e. the operation will complete before the function
+	 * returns, though the operation may fail.
+	 */
+	ONOFF_CLIENT_NOTIFY_NO_WAIT     = 0,
+
+	/* Indicates operation completion notification is through a
+	 * k_poll_signal reference.
+	 */
+	ONOFF_CLIENT_NOTIFY_SIGNAL      = 1,
+
+	/* Indicates operation completion notification is through a
+	 * callback.
+	 */
+	ONOFF_CLIENT_NOTIFY_CALLBACK    = 2,
+
+	/* This and higher bits reserved for internal use. */
+	ONOFF_CLIENT_INTERNAL_BASE      = BIT(2),
+};
+
+/* Forward declaration */
+struct onoff_client;
+
+/**
+ * @brief Signature used to notify an on-off service client of the
+ * completion of a transition.
+ *
+ * These functions may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-callable and non-suspendable.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param res the result of the transition.  Expected values are
+ * service-specific, but the value shall be non-negative if the
+ * operation succeeded, and negative if the operation failed.
+ */
+typedef void (*onoff_client_callback)(struct onoff_service *srv,
+				      struct onoff_client *cli,
+				      int res);
+
+/**
+ * @brief State associated with a client of an on-off service.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. onoff_client_init_signal()) to
+ * configure them.
+ *
+ * Control of the object content transfers to the service provider
+ * when a pointer to the object is passed to any on-off service
+ * function.  While the service provider controls the object the
+ * client must not change any object fields.  Control reverts to the
+ * client:
+ * * if the call to the service API returns an error; or
+ * * if the call to the service API succeeds for a no-wait operation; otherwise
+ * * when operation completion is posted (signalled or callback invoked).
+ *
+ * Only the result field is intended for use by clients, and it is
+ * available for inspection only after control reverts to the client.
+ */
+struct onoff_client {
+	/* Links the client into the set of waiting service users. */
+	sys_snode_t node;
+
+	union {
+		/* Pointer to signal used to notify client.
+		 *
+		 * The signal value corresponds to the res parameter
+		 * of onoff_client_callback.
+		 */
+		struct k_poll_signal *signal;
+
+		/* Pointer to callback used to notify client. */
+		onoff_client_callback callback;
+	} async;
+
+	/**
+	 * @brief The result of the transition operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for no-wait synchronous transitions.
+	 */
+	int result;
```

Comment:
Possibly.  It's useful for callback because the callback handler doesn't have to store the result somewhere for follow-on processing.  I'll look into moving it.

Reply:
may a service should store the result. It should be common for all clients since it comes from `onoff_service_notify_fn`

Reply:
It can't be held by the service because it could be overwritten at any time by another transition initiated by another thread/ISR.  Clients that were delayed in looking for it would get the wrong result.